### ISSUE TITLE
sql: enable changing the current user of a session with SET ROLE

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#15005]
Add support for the [`SET ROLE`](https://www.postgresql.org/docs/13/static/sql-set-role.html) command and variants from PostgreSQL. This command sets or resets the user for the current session.

This should also support `SET LOCAL ROLE`, which changes the role just for the current transaction. This is related to, but not dependent on `SET LOCAL` for session variables: #32562

At least one client, Flyway, requires support for this syntax. Graphile uses this too.

Epic CRDB-7217

**Comment by bdarnell** [comment]
It looks like flyway only requires [RESET ROLE](https://github.com/flyway/flyway/blob/053ad59be339a565a1e9ecbfaa6b9d3e0678a7d9/flyway-core/src/main/java/org/flywaydb/core/internal/dbsupport/postgresql/PostgreSQLDbSupport.java#L90), which is a trivial noop if we don't support SET ROLE.

**Comment by dianasaur323** [comment]
Looks like this is RBAC, which we will work on in 1.2. cc @nstewart

**Comment by nstewart** [comment]
RBAC is scheduled to be an enterprise feature, so we should definitely try to decouple that effort from the apparently small amount of work required to unblock FlywayDB, which should work for all users. I'll note this consideration in our RBAC record. cc @awoods187

**Comment by jordanlewis** [comment]
FWIW this isn't a blocker for FlywayDB anymore, since their team created an adapter that works around our limitations.

**Comment by cfsnyder** [comment]
@jordanlewis do you have a link for information about the Flyway adapter that you mentioned?

**Comment by jordanlewis** [comment]
@cfsnyder it'll be part of their next major release. Adapter is probably the wrong word - it'll be part of the main Flyway product.

**Comment by mberhault** [comment]
`SET ROLE` is not part of the [initial RBAC](https://github.com/cockroachdb/cockroach/pull/20149) implementation (users cannot specify that they want to run as a specific role. Not entirely needed as privileges are currently always inherited. SET ROLE may be needed once `NOINHERIT` is possible).

Since this is no longer needed for Flyway, I propose we close it.

**Comment by awoods187** [comment]
Moving to later in case we need it, but agree with Marc that its not needed at the moment cc @vivekmenezes

**Comment by mberhault** [comment]
Unassigning. This need to be prioritized. If desired, this would be an easy project for someone on the sql team.

**Comment by aidan-plenert-macdonald** [comment]
Can I get guidance on how to add  support for this feature?

**Comment by otan** [comment]
Unfortunately, due to the way we parse `SESSION_USER`:  https://github.com/cockroachdb/cockroach/blob/3ea1633cd0f353eecb0612df1c29841eaa94806e/pkg/sql/parser/sql.y#L11530, we may need two releases to push this out and work as expected. Since `SESSION_USER` and `CURRENT_USER` is different here, changing the builtin it parses to would not work in a mixed version state as any new builtin we introduce needs on release to make it in.

**Comment by otan** [comment]
#68973 has this working, but with the above limitation.
